<a href="https://colab.research.google.com/github/LeonardoGoncRibeiro/01_DataScienceUsingPython/blob/main/03_FunctionsPackagesAndPandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In this file, we will continue to show Python functionalities, as well as built-in functions and packages that may assist in Data Science.

In the end, we will show Pandas library, another very strong tool for handling data. 

In [ ]:
!python -V

Python 3.7.13


In [ ]:
import pandas as pd                            # Importing pandas
# pd.set_option('display.max_rows', 1000)      # If you want pandas to display all lines, uncomment this line of code

dataset = pd.read_csv('db.csv', sep = ';')     # Reading csv

In [ ]:
dataset

Nome                  Motor   Ano  Quilometragem  \
0            Jetta Variant        Motor 4.0 Turbo  2003        44410.0   
1                   Passat           Motor Diesel  1991         5712.0   
2                 Crossfox        Motor Diesel V8  1990        37123.0   
3                      DS5        Motor 2.4 Turbo  2019            NaN   
4         Aston Martin DB4        Motor 2.4 Turbo  2006        25757.0   
..                     ...                    ...   ...            ...   
253           Phantom 2013               Motor V8  2014        27505.0   
254  Cadillac Ciel concept               Motor V8  1991        29981.0   
255             Classe GLK  Motor 5.0 V8 Bi-Turbo  2002        52637.0   
256       Aston Martin DB5           Motor Diesel  1996         7685.0   
257                  Macan        Motor Diesel V6  1992        50188.0   

     Zero_km                                         Acessórios      Valor  
0      False  ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
1      False  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
2      False  ['Piloto automático', 'Controle de estabilidad...   72832.16  
3       True  ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
4      False  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10  
..       ...                                                ...        ...  
253    False  ['Controle de estabilidade', 'Piloto automátic...   51759.58  
254    False  ['Bancos de couro', 'Painel digital', 'Sensor ...   51667.06  
255    False  ['Rodas de liga', 'Controle de tração', 'Câmbi...   68934.03  
256    False  ['Ar condicionado', '4 X 4', 'Câmbio automátic...  122110.90  
257    False  ['Central multimídia', 'Teto panorâmico', 'Vid...   90381.47  

[258 rows x 7 columns]

We can see the data type for a Pandas dataframe.

In [ ]:
dataset.dtypes

Nome              object
Motor             object
Ano                int64
Quilometragem    float64
Zero_km             bool
Acessórios        object
Valor            float64
dtype: object

Note that, in a Pandas data frame, we may have different data types in the same dataset.

We may use the describe( ) method to get some basic statistics for some variables

In [ ]:
dataset[['Quilometragem', 'Valor']].describe( )

Quilometragem          Valor
count     197.000000     258.000000
mean    58278.421320   98960.513101
std     35836.733259   29811.932305
min       107.000000   50742.100000
25%     27505.000000   70743.512500
50%     55083.000000   97724.380000
75%     90495.000000  124633.302500
max    119945.000000  149489.920000

Also, we can get some information from our dataset using the info( ) method

In [ ]:
dataset.info( )

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 258 entries, 0 to 257
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Nome           258 non-null    object 
 1   Motor          258 non-null    object 
 2   Ano            258 non-null    int64  
 3   Quilometragem  197 non-null    float64
 4   Zero_km        258 non-null    bool   
 5   Acessórios     258 non-null    object 
 6   Valor          258 non-null    float64
dtypes: bool(1), float64(2), int64(1), object(3)
memory usage: 12.5+ KB


In [ ]:
%reset-f

# Tuples

Tuples are very similar to lists. However, they are not mutable.

Tuples are commonly used to store a collection of values, usually non-homogeneous.

Tuples can be built in a variety of ways.

In [ ]:
( )

()

In [ ]:
1, 2, 3

(1, 2, 3)

In [ ]:
name = 'Passat'
value = 153000

(name, value)

('Passat', 153000)

In [ ]:
cars = ['Jetta', 'Passat', 'Crossfox', 'DS5']

cars_tuple = tuple(cars)

In [ ]:
type(cars_tuple)

tuple

# Selection in tuples

Selection in tuples is very similar to indexing in lists

In [ ]:
cars_tuple[0]

'Jetta'

In [ ]:
cars_tuple[2]

'Crossfox'

In [ ]:
cars_tuple[-1]

'DS5'

In [ ]:
cars_tuple[1:3]

('Passat', 'Crossfox')

In [ ]:
cars_tuple[:2]

('Jetta', 'Passat')

# Iterations using tuples

Again, iterating in tuples is very similar to iterating in lists

In [ ]:
for item in cars_tuple:
  print(item)

Jetta
Passat
Crossfox
DS5


We can assign values to multiple variables from a single tuple

In [ ]:
car_1, car_2, car_3, car_4 = cars_tuple

In [ ]:
print(car_1)
print(car_2)
print(car_3)
print(car_4)

Jetta
Passat
Crossfox
DS5


You may use _ to assist

In [ ]:
_, car_A, _, car_B = cars_tuple

In [ ]:
print(car_A)
print(car_B)

Passat
DS5


In [ ]:
_, car_Passat, *_ = cars_tuple

In [ ]:
print(car_Passat)

Passat


Iterations using tuples can be performed using the zip() function

In [ ]:
%reset-f

In [ ]:
cars = ['Jetta', 'Passat', 'Crossfox', 'DS5']
values = [88078.64, 106161.94, 72832.16, 124549.07]

In [ ]:
list(zip(cars, values))

[('Jetta', 88078.64),
 ('Passat', 106161.94),
 ('Crossfox', 72832.16),
 ('DS5', 124549.07)]

In [ ]:
for car in zip(cars, values):
  print(car)

('Jetta', 88078.64)
('Passat', 106161.94)
('Crossfox', 72832.16)
('DS5', 124549.07)


In [ ]:
for name, value in zip(cars, values):
  print(f"The price of car {name} is USD {value}.")

The price of car Jetta is USD 88078.64.
The price of car Passat is USD 106161.94.
The price of car Crossfox is USD 72832.16.
The price of car DS5 is USD 124549.07.


In [ ]:
%reset-f

# Dictionaries

Dictionaries are another structure that represents a collection of data. However, here, data has the same mapping type, which helps in the association of these values.

The first element is known as the *key*, and the second is known as the *value*

In [ ]:
cars = ['Jetta', 'Passat', 'Crossfox', 'DS5']
values = [88078.64, 106161.94, 72832.16, 124549.07]

In [ ]:
cars_dict = {'Jetta': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16, 'DS5':124549.07}
cars_dict

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta': 88078.64,
 'Passat': 106161.94}

In [ ]:
type(cars_dict)

dict

In [ ]:
%reset-f

We may use the zip( ) function to create dictionaries

In [ ]:
cars = ['Jetta', 'Passat', 'Crossfox', 'DS5']
values = [88078.64, 106161.94, 72832.16, 124549.07]
cars_dict = dict(zip(cars, values))

cars_dict

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta': 88078.64,
 'Passat': 106161.94}

# Operations using dictionaries

Unlike tuples, dictionaries are mutable. Thus, there are multiple operations that can be performed to change or to assess its values.

We may use the *key* to perform indexing on a dictionary

In [ ]:
cars_dict['Passat']

106161.94

We may also use this type of indexing to add new values to the dictionary

In [ ]:
cars_dict['Palio'] = 78920.20
cars_dict

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta': 88078.64,
 'Palio': 78920.2,
 'Passat': 106161.94}

To delete values, we may use del( )

In [ ]:
del cars_dict['Palio']
cars_dict

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta': 88078.64,
 'Passat': 106161.94}

Also, we may check if a given *key* is in our dictionary

In [ ]:
'Passat' in cars_dict

True

We may assess the length of a dictionary using len( )

In [ ]:
len(cars_dict)

4

Note that, as in Python lists, we may have dictionaries *inside* other dictionaries

In [ ]:
cars_data = {
    'Passat': {
        'ano': 2012,
        'km': 50000,
        'valor': 75000,
        'acessorios': ['Airbag', 'ABS']
    }, 
    'Crossfox': {
        'ano': 2015,
        'km': 35000,
        'valor': 25000
    }
}

In [ ]:
'acessorios' in cars_data['Crossfox']

False

In [ ]:
'acessorios' in cars_data['Passat']

True

In [ ]:
cars_data['Crossfox']['valor']

25000

In [ ]:
cars_data['Passat']['acessorios'][-1]

'ABS'

# Methods for dictionaries

update( ) allows us to update our dictionary

In [ ]:
cars_dict

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta': 88078.64,
 'Passat': 106161.94}

In [ ]:
cars_dict.update({'Palio': 78920.20})
cars_dict

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Jetta': 88078.64,
 'Palio': 78920.2,
 'Passat': 106161.94}

Update can also be used to change values from our dictionary

In [ ]:
cars_dict.update({'Crossfox': 74370.91, 'Aston Martin': 237110.23})
cars_dict

{'Aston Martin': 237110.23,
 'Crossfox': 74370.91,
 'DS5': 124549.07,
 'Jetta': 88078.64,
 'Palio': 78920.2,
 'Passat': 106161.94}

Similar to lists, we can use copy( ) for dictionaries

In [ ]:
copy_cars_dict = cars_dict.copy( )

We can use pop( ) to remove a *key* and its value from the dictionary

In [ ]:
copy_cars_dict.pop('DS5', 'Key not found') # 'Key not found is printed if key 'DS5' is not found

124549.07

In [ ]:
copy_cars_dict

{'Aston Martin': 237110.23,
 'Crossfox': 74370.91,
 'Jetta': 88078.64,
 'Palio': 78920.2,
 'Passat': 106161.94}

To clean the entire dictionary, we may use clear( )

In [ ]:
copy_cars_dict.clear( )

In [ ]:
copy_cars_dict

{}

# Iterations using dictionaries

keys( ) can be used to get a list of all keys from a dictionary

In [ ]:
cars_dict

{'Aston Martin': 237110.23,
 'Crossfox': 74370.91,
 'DS5': 124549.07,
 'Jetta': 88078.64,
 'Palio': 78920.2,
 'Passat': 106161.94}

In [ ]:
cars_dict.keys( )

dict_keys(['Jetta', 'Passat', 'Crossfox', 'DS5', 'Palio', 'Aston Martin'])

We can use keys( ) to help us in iterating in a dictionary

In [ ]:
for key in cars_dict.keys( ):
  print(cars_dict[key])

88078.64
106161.94
74370.91
124549.07
78920.2
237110.23


Similarly, we may use values( ) to get a list with all values from the dictionary

In [ ]:
cars_dict.values( )

dict_values([88078.64, 106161.94, 74370.91, 124549.07, 78920.2, 237110.23])

items( ) returns a list with a tuple for each key-value pair

In [ ]:
cars_dict.items( )

dict_items([('Jetta', 88078.64), ('Passat', 106161.94), ('Crossfox', 74370.91), ('DS5', 124549.07), ('Palio', 78920.2), ('Aston Martin', 237110.23)])

In [ ]:
for car, v in cars_dict.items( ):
  print(f'{car} costs USD {v}')

Jetta costs USD 88078.64
Passat costs USD 106161.94
Crossfox costs USD 74370.91
DS5 costs USD 124549.07
Palio costs USD 78920.2
Aston Martin costs USD 237110.23


In [ ]:
%reset-f

# Python built-in functions

Functions are a re-usable unit of code that can be used to perform tasks

Different functions were already here, such as type( ), print( ), zip( ), len( ), set( ), and others.

https://docs.python.org/3/library/functions.html

In [ ]:
cars = ['Jetta', 'Passat', 'Crossfox', 'DS5']
values = [88078.64, 106161.94, 72832.16, 124549.07]
cars_dict = dict(zip(cars, values))

For instance, if we want to create a list with all car values, and return their sum, we can use:

In [ ]:
val = []
s = 0

for v in cars_dict.values( ):
  val.append(v)
  s += v

print(s)

391621.81


However, we can perform the same procedure using a built-in function

In [ ]:
s = sum(cars_dict.values())
s

391621.81

We can use help( ) to print documentation about a given built-in function

In [ ]:
help(sum)

Help on built-in function sum in module builtins:

sum(iterable, start=0, /)
    Return the sum of a 'start' value (default: 0) plus an iterable of numbers
    
    When the iterable is empty, return the start value.
    This function is intended specifically for use with numeric values and may
    reject non-numeric types.



If there are no built-in functions for a specific problem, we may create our own functions.

```
def <func_name>():
  <func_instructions>
```



For instance, if we want to create a function to evaluate the mean

In [ ]:
def eval_mean_np( ):
  values = [1, 3, 5, 6]
  print(sum(values)/len(values))

In [ ]:
eval_mean_np( )

3.75


To guarantee that functions can be used in different parts of the code and with different values, we may want to use parameters

In [ ]:
def eval_mean(values):
  print(sum(values)/len(values))

In [ ]:
eval_mean([23, 45, 67])

45.0


We can create our function to handle different types of data

In [ ]:
cars_data = {
    'Crossfox': {'km': 35000, 'ano': 2005}, 
    'DS5': {'km': 17000, 'ano': 2015}, 
    'Fusca': {'km': 130000, 'ano': 1979}, 
    'Jetta': {'km': 56000, 'ano': 2011}, 
    'Passat': {'km': 62000, 'ano': 1999}
}

In [ ]:
cars_data.items()

dict_items([('Crossfox', {'km': 35000, 'ano': 2005}), ('DS5', {'km': 17000, 'ano': 2015}), ('Fusca', {'km': 130000, 'ano': 1979}), ('Jetta', {'km': 56000, 'ano': 2011}), ('Passat', {'km': 62000, 'ano': 1999})])

In [ ]:
def calc_km_per_year(data, curr_year):
  for car in data.items( ):
     km_per_year = car[1]['km']/(curr_year - car[1]['ano'])
     print(f'{car[0]}: {km_per_year} km per year')

In [ ]:
current_year = 2022
calc_km_per_year(cars_data, current_year)

Crossfox: 2058.823529411765 km per year
DS5: 2428.5714285714284 km per year
Fusca: 3023.2558139534885 km per year
Jetta: 5090.909090909091 km per year
Passat: 2695.6521739130435 km per year


We can also return outputs from our user-defined function

In [ ]:
def eval_mean(values):
  return (sum(values)/len(values))

*return* makes it so that the function will return a given value

In [ ]:
mean_value = eval_mean([62, 54, 92])

In [ ]:
mean_value

69.33333333333333

We may return multiple values in the same function using lists

In [ ]:
def eval_mean_len(values):
  return [(sum(values)/len(values)), len(values)]

In [ ]:
descr = eval_mean_len([32, 33, 92])

In [ ]:
descr

[52.333333333333336, 3]

In [ ]:
%reset-f

# Pandas library

Finally, we will continue to use Pandas.

In Pandas, data is stored in specific structures: Series and DataFrames. 

DataFrames are similar to a dictionary, but they make it is easier to handle and visualize data.

A DataFrame can be created from a dicitionary, a list, a series, a constant...


In [1]:
import pandas as pd

A Pandas Series can be created from a list

In [10]:
cars = ['Jetta', 'Passat', 'Crossfox', 'DS5']
values = [88078.64, 106161.94, 72832.16, 124549.07]

In [11]:
pd.Series(cars) # Creating a series from a list

0       Jetta
1      Passat
2    Crossfox
3         DS5
dtype: object

A Pandas DataFrame can be created, for instance, from a list of dictionaries

In [12]:
data = [
        {'name': 'Jetta', 'engine': '4.0 Turbo', 'year': 2003, 'km': 44410.0, 'zero_km': False, 'value': 88078.64},
        {'name': 'Passat', 'engine': 'Diesel', 'year': 1991, 'km': 5712.0, 'zero_km': False, 'value': 106161.94},
        {'name': 'Crossfox', 'engine': 'Diesel V8', 'year': 1990, 'km': 37123.0, 'zero_km': False, 'value': 72823.16},
] 

In [13]:
dataset = pd.DataFrame(data)

In [14]:
dataset

name     engine  year       km  zero_km      value
0     Jetta  4.0 Turbo  2003  44410.0    False   88078.64
1    Passat     Diesel  1991   5712.0    False  106161.94
2  Crossfox  Diesel V8  1990  37123.0    False   72823.16

Note that dictionary keys are used as the header for each column. We can change the order of columns

In [15]:
dataset[['engine', 'value', 'km', 'name', 'year', 'zero_km']]

engine      value       km      name  year  zero_km
0  4.0 Turbo   88078.64  44410.0     Jetta  2003    False
1     Diesel  106161.94   5712.0    Passat  1991    False
2  Diesel V8   72823.16  37123.0  Crossfox  1990    False

Usually, DataFrames are actually created via an external file.

In [18]:
pd.read_csv('db.csv', sep = ';')

Nome                  Motor   Ano  Quilometragem  \
0            Jetta Variant        Motor 4.0 Turbo  2003        44410.0   
1                   Passat           Motor Diesel  1991         5712.0   
2                 Crossfox        Motor Diesel V8  1990        37123.0   
3                      DS5        Motor 2.4 Turbo  2019            NaN   
4         Aston Martin DB4        Motor 2.4 Turbo  2006        25757.0   
..                     ...                    ...   ...            ...   
253           Phantom 2013               Motor V8  2014        27505.0   
254  Cadillac Ciel concept               Motor V8  1991        29981.0   
255             Classe GLK  Motor 5.0 V8 Bi-Turbo  2002        52637.0   
256       Aston Martin DB5           Motor Diesel  1996         7685.0   
257                  Macan        Motor Diesel V6  1992        50188.0   

     Zero_km                                         Acessórios      Valor  
0      False  ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
1      False  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
2      False  ['Piloto automático', 'Controle de estabilidad...   72832.16  
3       True  ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
4      False  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10  
..       ...                                                ...        ...  
253    False  ['Controle de estabilidade', 'Piloto automátic...   51759.58  
254    False  ['Bancos de couro', 'Painel digital', 'Sensor ...   51667.06  
255    False  ['Rodas de liga', 'Controle de tração', 'Câmbi...   68934.03  
256    False  ['Ar condicionado', '4 X 4', 'Câmbio automátic...  122110.90  
257    False  ['Central multimídia', 'Teto panorâmico', 'Vid...   90381.47  

[258 rows x 7 columns]

# Indexing using DataFrames

In [21]:
dataset = pd.read_csv('db.csv', sep = ';', index_col = 0)

DataFrame indexing can be performed using the column titles

In [22]:
dataset['Valor']

Nome
Jetta Variant             88078.64
Passat                   106161.94
Crossfox                  72832.16
DS5                      124549.07
Aston Martin DB4          92612.10
                           ...    
Phantom 2013              51759.58
Cadillac Ciel concept     51667.06
Classe GLK                68934.03
Aston Martin DB5         122110.90
Macan                     90381.47
Name: Valor, Length: 258, dtype: float64

Note that the data type of one DataFrame column is a Series

In [24]:
type(dataset['Valor'])

pandas.core.series.Series

Usual rules for splicing lists and tuples also work with Pandas DataFrame

In [28]:
dataset[2:4]

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                      
Crossfox  Motor Diesel V8  1990        37123.0    False   
DS5       Motor 2.4 Turbo  2019            NaN     True   

                                                 Acessórios      Valor  
Nome                                                                    
Crossfox  ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5       ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07

Using .loc, one is able to select a group of row/columns

In [29]:
dataset.loc['Passat'] 

Motor                                                 Motor Diesel
Ano                                                           1991
Quilometragem                                               5712.0
Zero_km                                                      False
Acessórios       ['Central multimídia', 'Teto panorâmico', 'Fre...
Valor                                                    106161.94
Name: Passat, dtype: object

In [32]:
dataset.loc[['Passat', 'DS5']]

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                    
Passat     Motor Diesel  1991         5712.0    False   
DS5     Motor 2.4 Turbo  2019            NaN     True   

                                               Acessórios      Valor  
Nome                                                                  
Passat  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
DS5     ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07

In [34]:
dataset.loc[['Passat', 'DS5'], ['Ano', 'Valor']]

Ano      Valor
Nome                   
Passat  1991  106161.94
DS5     2019  124549.07

In [37]:
dataset.loc[:, ['Ano', 'Valor']]

Ano      Valor
Nome                                  
Jetta Variant          2003   88078.64
Passat                 1991  106161.94
Crossfox               1990   72832.16
DS5                    2019  124549.07
Aston Martin DB4       2006   92612.10
...                     ...        ...
Phantom 2013           2014   51759.58
Cadillac Ciel concept  1991   51667.06
Classe GLK             2002   68934.03
Aston Martin DB5       1996  122110.90
Macan                  1992   90381.47

[258 rows x 2 columns]

Indexing with numerical values can be performed using .iloc

In [40]:
dataset.iloc[3]

Motor                                              Motor 2.4 Turbo
Ano                                                           2019
Quilometragem                                                  NaN
Zero_km                                                       True
Acessórios       ['Travas elétricas', '4 X 4', 'Vidros elétrico...
Valor                                                    124549.07
Name: DS5, dtype: object

In [41]:
dataset.iloc[4:9, [0, 5, 2]]


Motor      Valor  Quilometragem
Nome                                                       
Aston Martin DB4  Motor 2.4 Turbo   92612.10        25757.0
Palio Weekend       Motor 1.8 16v   97497.73        10728.0
A5                Motor 4.0 Turbo   56445.20            NaN
Série 3 Cabrio       Motor 1.0 8v  112310.44        77599.0
Dodge Jorney        Motor 3.0 32v  120716.27        99197.0

# Queries using DataFrames

In [43]:
dataset.head( )

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios      Valor  
Nome                                                                            
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox          ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10

In [50]:
dataset['Motor'] 

Nome
Jetta Variant                  Motor 4.0 Turbo
Passat                            Motor Diesel
Crossfox                       Motor Diesel V8
DS5                            Motor 2.4 Turbo
Aston Martin DB4               Motor 2.4 Turbo
                                 ...          
Phantom 2013                          Motor V8
Cadillac Ciel concept                 Motor V8
Classe GLK               Motor 5.0 V8 Bi-Turbo
Aston Martin DB5                  Motor Diesel
Macan                          Motor Diesel V6
Name: Motor, Length: 258, dtype: object

In [48]:
dataset.Motor == 'Motor Diesel'

Nome
Jetta Variant            False
Passat                    True
Crossfox                 False
DS5                      False
Aston Martin DB4         False
                         ...  
Phantom 2013             False
Cadillac Ciel concept    False
Classe GLK               False
Aston Martin DB5          True
Macan                    False
Name: Motor, Length: 258, dtype: bool

In [47]:
dataset[dataset.Motor == 'Motor Diesel']

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                
Passat                 Motor Diesel  1991         5712.0    False   
Effa Hafei Picape Baú  Motor Diesel  1991       102959.0    False   
Sorento                Motor Diesel  2019            NaN     True   
New Fiesta Hatch       Motor Diesel  2017       118895.0    False   
Kangoo Express         Motor Diesel  2007        29132.0    False   
Fit                    Motor Diesel  2013        44329.0    False   
Cielo Hatch            Motor Diesel  2019            NaN     True   
Symbol                 Motor Diesel  2016       117714.0    False   
A4 Sedan               Motor Diesel  2002        30511.0    False   
A4 Avant               Motor Diesel  2014        17357.0    False   
Silver Shadow          Motor Diesel  2015        99052.0    False   
Camry                  Motor Diesel  2019            NaN     True   
Clio                   Motor Diesel  1990        66437.0    False   
Série 3 M3 Coupé       Motor Diesel  2013        19896.0    False   
Gallardo LP 560 – 4    Motor Diesel  2000       113045.0    False   
Chana Utility          Motor Diesel  2009        52034.0    False   
TT Roadster            Motor Diesel  1998       107392.0    False   
Aston Martin Virage    Motor Diesel  2019            NaN     True   
Sandero                Motor Diesel  2015        34783.0    False   
XKR                    Motor Diesel  1996        85127.0    False   
X5                     Motor Diesel  2002        13606.0    False   
Dodge Dakota           Motor Diesel  1993        71544.0    False   
i30 CW                 Motor Diesel  2015         8497.0    False   
Série 7 Sedã           Motor Diesel  2019            NaN     True   
V60                    Motor Diesel  2004        91840.0    False   
Aston Martin DB5       Motor Diesel  1996         7685.0    False   

                                                              Acessórios  \
Nome                                                                       
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Effa Hafei Picape Baú  ['Controle de estabilidade', 'Painel digital',...   
Sorento                ['Sensor de chuva', 'Câmera de estacionamento'...   
New Fiesta Hatch       ['Sensor de estacionamento', 'Travas elétricas...   
Kangoo Express         ['Bancos de couro', 'Câmbio automático', 'Pilo...   
Fit                    ['Freios ABS', 'Câmera de estacionamento', 'Câ...   
Cielo Hatch            ['Painel digital', 'Central multimídia', 'Câme...   
Symbol                 ['4 X 4', 'Piloto automático', 'Sensor crepusc...   
A4 Sedan               ['Câmera de estacionamento', '4 X 4', 'Travas ...   
A4 Avant               ['Teto panorâmico', '4 X 4', 'Bancos de couro'...   
Silver Shadow          ['4 X 4', 'Central multimídia', 'Ar condiciona...   
Camry                  ['Travas elétricas', 'Rodas de liga', 'Sensor ...   
Clio                   ['4 X 4', 'Câmbio automático', 'Freios ABS', '...   
Série 3 M3 Coupé       ['Bancos de couro', 'Sensor de chuva', 'Ar con...   
Gallardo LP 560 – 4    ['Freios ABS', 'Teto panorâmico', 'Central mul...   
Chana Utility          ['Ar condicionado', 'Sensor de estacionamento'...   
TT Roadster            ['Ar condicionado', 'Sensor de estacionamento'...   
Aston Martin Virage    ['Travas elétricas', 'Controle de tração', 'Câ...   
Sandero                ['Piloto automático', 'Teto panorâmico', 'Câmb...   
XKR                    ['Controle de tração', 'Ar condicionado', 'Tra...   
X5                     ['Controle de tração', 'Piloto automático', 'T...   
Dodge Dakota           ['Controle de tração', 'Sensor de estacionamen...   
i30 CW                 ['Sensor de chuva', 'Vidros elétricos', 'Contr...   
Série 7 Sedã           ['Vidros elétricos', 'Travas elétricas', 'Roda...   
V60                    ['Rodas de liga', 'Sensor crepuscular', 'Ar co...   
Aston Martin DB5       ['Ar condicionado', '4

This boolean indexing returns every car where 'Motor' == 'Motor Diesel'

We can use multiple conditions in this query

In [61]:
sel1 = dataset.Motor == 'Motor Diesel'
sel2 = dataset.Zero_km == True

In [62]:
sel1

Nome
Jetta Variant            False
Passat                    True
Crossfox                 False
DS5                      False
Aston Martin DB4         False
                         ...  
Phantom 2013             False
Cadillac Ciel concept    False
Classe GLK               False
Aston Martin DB5          True
Macan                    False
Name: Motor, Length: 258, dtype: bool

In [63]:
sel2

Nome
Jetta Variant            False
Passat                   False
Crossfox                 False
DS5                       True
Aston Martin DB4         False
                         ...  
Phantom 2013             False
Cadillac Ciel concept    False
Classe GLK               False
Aston Martin DB5         False
Macan                    False
Name: Zero_km, Length: 258, dtype: bool

In [64]:
dataset[sel1 & sel2] # & and, | or

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Sorento              Motor Diesel  2019            NaN     True   
Cielo Hatch          Motor Diesel  2019            NaN     True   
Camry                Motor Diesel  2019            NaN     True   
Aston Martin Virage  Motor Diesel  2019            NaN     True   
Série 7 Sedã         Motor Diesel  2019            NaN     True   

                                                            Acessórios  \
Nome                                                                     
Sorento              ['Sensor de chuva', 'Câmera de estacionamento'...   
Cielo Hatch          ['Painel digital', 'Central multimídia', 'Câme...   
Camry                ['Travas elétricas', 'Rodas de liga', 'Sensor ...   
Aston Martin Virage  ['Travas elétricas', 'Controle de tração', 'Câ...   
Série 7 Sedã         ['Vidros elétricos', 'Travas elétricas', 'Roda...   

                         Valor  
Nome                            
Sorento               81399.35  
Cielo Hatch          145197.70  
Camry                138597.27  
Aston Martin Virage   97290.18  
Série 7 Sedã          67539.79

We can also use the query method

In [67]:
dataset.query('Motor == "Motor Diesel" and Zero_km == True')

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Sorento              Motor Diesel  2019            NaN     True   
Cielo Hatch          Motor Diesel  2019            NaN     True   
Camry                Motor Diesel  2019            NaN     True   
Aston Martin Virage  Motor Diesel  2019            NaN     True   
Série 7 Sedã         Motor Diesel  2019            NaN     True   

                                                            Acessórios  \
Nome                                                                     
Sorento              ['Sensor de chuva', 'Câmera de estacionamento'...   
Cielo Hatch          ['Painel digital', 'Central multimídia', 'Câme...   
Camry                ['Travas elétricas', 'Rodas de liga', 'Sensor ...   
Aston Martin Virage  ['Travas elétricas', 'Controle de tração', 'Câ...   
Série 7 Sedã         ['Vidros elétricos', 'Travas elétricas', 'Roda...   

                         Valor  
Nome                            
Sorento               81399.35  
Cielo Hatch          145197.70  
Camry                138597.27  
Aston Martin Virage   97290.18  
Série 7 Sedã          67539.79

# Iterating using DataFrames

DataFrames can not be iterated like lists or tuples

In [69]:
for item in dataset:
  print(item)

Motor
Ano
Quilometragem
Zero_km
Acessórios
Valor


In practice, we have to use an specific method: iterrows( )

In [73]:
list(dataset.iterrows())

[('Jetta Variant',
  Motor                                              Motor 4.0 Turbo
  Ano                                                           2003
  Quilometragem                                              44410.0
  Zero_km                                                      False
  Acessórios       ['Rodas de liga', 'Travas elétricas', 'Piloto ...
  Valor                                                     88078.64
  Name: Jetta Variant, dtype: object),
 ('Passat', Motor                                                 Motor Diesel
  Ano                                                           1991
  Quilometragem                                               5712.0
  Zero_km                                                      False
  Acessórios       ['Central multimídia', 'Teto panorâmico', 'Fre...
  Valor                                                    106161.94
  Name: Passat, dtype: object),
 ('Crossfox',
  Motor                                              Motor

In [118]:
import math

curr_year = 2022

for index, row in dataset.iterrows( ):
  if (math.isnan(row['Quilometragem']) or row['Ano'] == curr_year):
    if (index == 'Sorento'):
      print(math.isnan(row['Quilometragem']))
      print('enter1')
    km_per_year = 0
  else:    
    if (index == 'Sorento'):
      print(math.isnan(row['Quilometragem']))
      print('enter2')
    km_per_year = row['Quilometragem']/(curr_year - row['Ano'])
  dataset.loc[index, 'KmPerYear'] = km_per_year
  

True
enter1


In [119]:
dataset.head( )

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios  \
Nome                                                                  
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox          ['Piloto automático', 'Controle de estabilidad...   
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   

                      Valor    KmPerYear  
Nome                                      
Jetta Variant      88078.64  2337.368421  
Passat            106161.94   184.258065  
Crossfox           72832.16  1160.093750  
DS5               124549.07     0.000000  
Aston Martin DB4   92612.10  1609.812500

Other ways of iterating using DataFrames

In [134]:
list_header = []

for item in dataset.head( ).items( ):
  list_header.append(item)

list_header

In [135]:
list_header = []

for item in dataset.head( ).itertuples( ):
  list_header.append(item)

list_header

[Pandas(Index='Jetta Variant', Motor='Motor 4.0 Turbo', Ano=2003, Quilometragem=44410.0, Zero_km=False, Acessórios="['Rodas de liga', 'Travas elétricas', 'Piloto automático', 'Bancos de couro', 'Ar condicionado', 'Sensor de estacionamento', 'Sensor crepuscular', 'Sensor de chuva']", Valor=88078.64, KmPerYear=2337.3684210526317),
 Pandas(Index='Passat', Motor='Motor Diesel', Ano=1991, Quilometragem=5712.0, Zero_km=False, Acessórios="['Central multimídia', 'Teto panorâmico', 'Freios ABS', '4 X 4', 'Painel digital', 'Piloto automático', 'Bancos de couro', 'Câmera de estacionamento']", Valor=106161.94, KmPerYear=184.25806451612902),
 Pandas(Index='Crossfox', Motor='Motor Diesel V8', Ano=1990, Quilometragem=37123.0, Zero_km=False, Acessórios="['Piloto automático', 'Controle de estabilidade', 'Sensor crepuscular', 'Freios ABS', 'Câmbio automático', 'Bancos de couro', 'Central multimídia', 'Vidros elétricos']", Valor=72832.16, KmPerYear=1160.09375),
 Pandas(Index='DS5', Motor='Motor 2.4 Turbo

# Data cleaning

Most of the time, data needs cleaning before being used for inferences and predictions.

info( ) is a very important method for understanding how a dataset is structured

In [136]:
dataset.info( )

<class 'pandas.core.frame.DataFrame'>
Index: 258 entries, Jetta Variant to Macan
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Motor          258 non-null    object 
 1   Ano            258 non-null    int64  
 2   Quilometragem  197 non-null    float64
 3   Zero_km        258 non-null    bool   
 4   Acessórios     258 non-null    object 
 5   Valor          258 non-null    float64
 6   KmPerYear      258 non-null    float64
dtypes: bool(1), float64(3), int64(1), object(2)
memory usage: 22.5+ KB


Note that there are 258 entries in this dataset. However, in index 2 ('Quilometragem') there are only 197 non-null values.

Let's understand why that is the case

In [142]:
sel = dataset['Quilometragem'].isna( )

In [143]:
dataset[sel]

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                  
DS5                   Motor 2.4 Turbo  2019            NaN     True   
A5                    Motor 4.0 Turbo  2019            NaN     True   
J5                           Motor V6  2019            NaN     True   
A3                       Motor 1.0 8v  2019            NaN     True   
Série 1 M                    Motor V8  2019            NaN     True   
...                               ...   ...            ...      ...   
Lamborghini Reventón  Motor 4.0 Turbo  2019            NaN     True   
Benni Mini                   Motor V8  2019            NaN     True   
Uno                   Motor Diesel V6  2019            NaN     True   
Santa Fe                Motor 3.0 32v  2019            NaN     True   
XC60                  Motor 4.0 Turbo  2019            NaN     True   

                                                             Acessórios  \
Nome                                                                      
DS5                   ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
A5                    ['Câmbio automático', 'Câmera de estacionament...   
J5                    ['Sensor crepuscular', 'Painel digital', 'Roda...   
A3                    ['4 X 4', 'Piloto automático', 'Central multim...   
Série 1 M             ['Controle de estabilidade', 'Central multimíd...   
...                                                                 ...   
Lamborghini Reventón  ['Controle de tração', 'Ar condicionado', 'Cen...   
Benni Mini            ['Sensor crepuscular', 'Câmbio automático', 'C...   
Uno                   ['Central multimídia', 'Sensor crepuscular', '...   
Santa Fe              ['Travas elétricas', 'Ar condicionado', '4 X 4...   
XC60                  ['Painel digital', 'Piloto automático', 'Centr...   

                          Valor  KmPerYear  
Nome                                        
DS5                   124549.07        0.0  
A5                     56445.20        0.0  
J5                     53183.38        0.0  
A3                     88552.39        0.0  
Série 1 M              94564.40        0.0  
...                         ...        ...  
Lamborghini Reventón   67664.86        0.0  
Benni Mini            126247.84        0.0  
Uno                   128852.21        0.0  
Santa Fe              129415.33        0.0  
XC60                   77675.79        0.0  

[61 rows x 7 columns]

Note that, when 'Quilometragem' == NaN, Zero_km == True.

That means that, here, NaN really means 0 (zero), as in: when the car is new (Zero_km == True), it has no "Quilometragem".

We can fill the missing values using the fillna( ) method

In [147]:
dataset.fillna(0, inplace = True)
dataset

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                         
Jetta Variant                Motor 4.0 Turbo  2003        44410.0    False   
Passat                          Motor Diesel  1991         5712.0    False   
Crossfox                     Motor Diesel V8  1990        37123.0    False   
DS5                          Motor 2.4 Turbo  2019            0.0     True   
Aston Martin DB4             Motor 2.4 Turbo  2006        25757.0    False   
...                                      ...   ...            ...      ...   
Phantom 2013                        Motor V8  2014        27505.0    False   
Cadillac Ciel concept               Motor V8  1991        29981.0    False   
Classe GLK             Motor 5.0 V8 Bi-Turbo  2002        52637.0    False   
Aston Martin DB5                Motor Diesel  1996         7685.0    False   
Macan                        Motor Diesel V6  1992        50188.0    False   

                                                              Acessórios  \
Nome                                                                       
Jetta Variant          ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox               ['Piloto automático', 'Controle de estabilidad...   
DS5                    ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4       ['Rodas de liga', '4 X 4', 'Central multimídia...   
...                                                                  ...   
Phantom 2013           ['Controle de estabilidade', 'Piloto automátic...   
Cadillac Ciel concept  ['Bancos de couro', 'Painel digital', 'Sensor ...   
Classe GLK             ['Rodas de liga', 'Controle de tração', 'Câmbi...   
Aston Martin DB5       ['Ar condicionado', '4 X 4', 'Câmbio automátic...   
Macan                  ['Central multimídia', 'Teto panorâmico', 'Vid...   

                           Valor    KmPerYear  
Nome                                           
Jetta Variant           88078.64  2337.368421  
Passat                 106161.94   184.258065  
Crossfox                72832.16  1160.093750  
DS5                    124549.07     0.000000  
Aston Martin DB4        92612.10  1609.812500  
...                          ...          ...  
Phantom 2013            51759.58  3438.125000  
Cadillac Ciel concept   51667.06   967.129032  
Classe GLK              68934.03  2631.850000  
Aston Martin DB5       122110.90   295.576923  
Macan                   90381.47  1672.933333  

[258 rows x 7 columns]

In [151]:
# dataset[dataset['Zero_km'] == True]
dataset.query('Zero_km == True')

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                  
DS5                   Motor 2.4 Turbo  2019            0.0     True   
A5                    Motor 4.0 Turbo  2019            0.0     True   
J5                           Motor V6  2019            0.0     True   
A3                       Motor 1.0 8v  2019            0.0     True   
Série 1 M                    Motor V8  2019            0.0     True   
...                               ...   ...            ...      ...   
Lamborghini Reventón  Motor 4.0 Turbo  2019            0.0     True   
Benni Mini                   Motor V8  2019            0.0     True   
Uno                   Motor Diesel V6  2019            0.0     True   
Santa Fe                Motor 3.0 32v  2019            0.0     True   
XC60                  Motor 4.0 Turbo  2019            0.0     True   

                                                             Acessórios  \
Nome                                                                      
DS5                   ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
A5                    ['Câmbio automático', 'Câmera de estacionament...   
J5                    ['Sensor crepuscular', 'Painel digital', 'Roda...   
A3                    ['4 X 4', 'Piloto automático', 'Central multim...   
Série 1 M             ['Controle de estabilidade', 'Central multimíd...   
...                                                                 ...   
Lamborghini Reventón  ['Controle de tração', 'Ar condicionado', 'Cen...   
Benni Mini            ['Sensor crepuscular', 'Câmbio automático', 'C...   
Uno                   ['Central multimídia', 'Sensor crepuscular', '...   
Santa Fe              ['Travas elétricas', 'Ar condicionado', '4 X 4...   
XC60                  ['Painel digital', 'Piloto automático', 'Centr...   

                          Valor  KmPerYear  
Nome                                        
DS5                   124549.07        0.0  
A5                     56445.20        0.0  
J5                     53183.38        0.0  
A3                     88552.39        0.0  
Série 1 M              94564.40        0.0  
...                         ...        ...  
Lamborghini Reventón   67664.86        0.0  
Benni Mini            126247.84        0.0  
Uno                   128852.21        0.0  
Santa Fe              129415.33        0.0  
XC60                   77675.79        0.0  

[61 rows x 7 columns]

In [155]:
%reset-f

In [156]:
import pandas as pd
dataset = pd.read_csv('db.csv', sep = ';')

Instead, we might want to actually remove those lines with missing values. In this case, we may use the method dropna( )

In [158]:
dataset.dropna(subset = ['Quilometragem'], inplace = True)